include/zephyr/net/tftp.h

```diff
@@ -53,41 +44,133 @@ struct tftpc {
 #define TFTPC_REMOTE_ERROR       -4
 #define TFTPC_RETRIES_EXHAUSTED  -5
 
-/* @brief This function gets "file" from the remote server.
+/**
+ * @brief TFTP Asynchronous Events notified to the application from the module
+ *        through the callback registered by the application.
+ */
+enum tftp_evt_type {
+	/** DATA event when data is received from remote server.
+	 *
+	 * @note DATA event structure contains payload data and size.
+	 */
+	TFTP_EVT_DATA,
+
+	/**  ERROR event when error is received from remote server.
+	 *
+	 * @note ERROR event structure contains error code and message.
+	 */
+	TFTP_EVT_ERROR
+};
+
+/** @brief Parameters for data event. */
+struct tftp_data_param {
+	uint8_t *data_ptr;         /**< Pointer to binary data. */
+	uint32_t len;              /**< Length of binary data. */
+};
+
+/** @brief Parameters for error event. */
+struct tftp_error_param {
+	char *msg;                 /**< Error message. */
+	int code;                  /**< Error code. */
+};
+
+/**
+ * @brief Defines event parameters notified along with asynchronous events
+ *        to the application.
+ */
+union tftp_evt_param {
+	/** Parameters accompanying TFTP_EVT_DATA event. */
+	struct tftp_data_param data;
+
+	/** Parameters accompanying TFTP_EVT_ERROR event. */
+	struct tftp_error_param error;
+};
+
+/** @brief Defines TFTP asynchronous event notified to the application. */
+struct tftp_evt {
+	/** Identifies the event. */
+	enum tftp_evt_type type;
+
+	/** Contains parameters (if any) accompanying the event. */
+	union tftp_evt_param param;
+};
+
+/**
+ * @typedef tftp_callback_t
+ *
+ * @brief Asynchronous event notification callback registered by the
+ *        application.
+ *
+ * @param[in] evt Event description along with result and associated
+ *                parameters (if any).
+ */
+typedef void (*tftp_callback_t)(const struct tftp_evt *evt);
+
+/**
+ * @brief TFTP client definition to maintain information relevant to the
+ *        client.
+ */
+struct tftpc {
+	/** Control Block that represents the remote server */
+	struct sockaddr *server;
+
+	/** Receive buffer used for packet in RX path. */
+	uint8_t rx_buf[TFTPC_MAX_BUF_SIZE];
+
+	/** Transmit buffer used for packet in TX path. */
+	uint8_t tx_buf[TFTPC_MAX_BUF_SIZE];
+
+	/** Event notification callback */
+	tftp_callback_t callback;
+
```

Comment:
The problem of mutex is that the same thread can lock the mutex multiple times, which cannot prevent the re-entry of one active GET or PUT transfer from the same client in the same thread.

Reply:
But is that really the case though? The API is not really asynchronous, the function call will block until the transfer is complete, or an error takes place. So there's no chance the API functions could be re-entered from the same thread.

Reply:
Oh yeah that's paradoxical. Let's do like this,
.socket handle out of client context (both APIs are blocking, means full socket lifetime)
.mutex in client context, two of them, allow simultaneous GET and PUT transfer from different threads

Reply:
I don't really think those two routines are mutually exclusive. Both use `send_request()` internally, which makes use of both TX and RX buffers. Other internal routines also make use of those buffers, so GET/PUT potentially still can interfere with each other. We should have a single mutex to guar d both routines.

Reply:
send_request() is fine as different sock will be used, but tx_buffer and rx_buffer has problem with simutaneous GET/PUT, as they are both used in either GET or PUT transfer.

Let's make it simple, use a single mutext to guard both, making simutaneous GET/PUT not possible.

Reply:
Sequential requests to GET and/or PUT from the same thread of a client is not possible due to the blocking nature of the APIs, in which case a mutex is not necessary. However multiple requests to GET and/or PUT from different of threads of a client is possible. Since both tx_buf and rx_buf must be reserved for one active GET or PUT transfer, we need to use a mutex to make the second request from a different thread to wait.

Reply:
> Sequential requests to GET and/or PUT from the same thread of a client is not possible due to the blocking nature of the APIs, in which case a mutex is not necessary. 

Maybe I used a wrong word (sequential), but I meant issuing the next GET/PUT request after the previous one has finished:
```
tftpc_get();
tftpc_get();
...
tftpc_put();
```
and in that case there is obviously no need for mutex.

> However multiple requests to GET and/or PUT from different threads of a client is possible. Since both tx_buf and rx_buf must be reserved for one active GET or PUT transfer, we need to use a mutex to make the second request from a different thread to wait.

Assuming that user really wants to have that behavior (though I think this is unlikely, as it is very "specific" reuse of TFTP client instance), this can be quickly implemented with:
```
static struct k_mutex lock;
static struct tftpc client;

static void thread_1()
{
    k_mutex_lock(&lock);
    client.server = SERVER_1; /* Optionally */
    tftpc_get(&client, ...);
    k_mutex_unlock(&lock);
}

static void thread_2()
{
    k_mutex_lock(&lock);
    client.server = SERVER_2; /* Optionally */
    tftpc_put(&client, ...);
    k_mutex_unlock(&lock);
}
```
We cannot do any better than this internally in `struct tftpc`, i.e. we cannot make parallel (in terms of data transfer) requests to work with single instance. All we can do is protect whole object. If that is the case, then it is better to do it externally.

Even if user wants to implement above use-case, there is still a single `tftp_callback_t callback;`. If an error occurs on `tftpc_get()` or `tftpc_put()` (scheduled by mulitple threads), then how should user figure out which thread received that error if there is a single callback and a single instance? Really, if user wants access from multiple threads, then he already knows better what he is doing and it is much better to add a mutex **externally** or use any other synchronization mechanism. If user does not know what he is doing, then we are not helping much with putting mutex internally.

There are still `callback` and `server` members of `struct tftpc` that user might or might not want to protect. With an external mutex it is possible to change those fields from application perspective in a thread-safe manner. With internal mutex, we are still not allowing to do so.

> Do you see any overhead to use a mutex in the API?

There is always some overhead, both in RAM and in CPU. It is 28 bytes compared to ~550 (assuming single buffer, which looks possible), which is 5%. Non-negligible IMO. Putting mutex internally "solves" some 1% of TFTP client use cases. So overall I don't think it is justified.

Reply:
> Even if user wants to implement above use-case, there is still a single tftp_callback_t callback;. If an error occurs on tftp_get() or tftp_put() (scheduled by mulitple threads), then how should user figure out which thread received that error if there is a single callback and a single instance? 

That's a good point. I agree, that perhaps we should simply document that if the API is about to be used from different  threads, unique context structure should be used.

Reply:
I get the point. The internal mutex can only protect the client structure within the code scope of the library, but it does not help if the contents are altered in caller application. Let's remove the internal mutex and add a comment in API document.

Thanks very much, @mniestroj. I've learnt a lot from your review.